File: demo/js/clock.mjs
//
//  clock.mjs - the iOS and Android demo clients
//
//  Copyright 2026 Saxon Herschel Nicholls
//
//  Stand-ins for the two React Native apps: until M2 puts @super-log/client
//  inside the real Expo builds, this runs the exact same client from Node -
//  one process per device, same topics, same origin shape - so the
//  four-stream demo has all four topics live today. Swapping the real app
//  in changes nothing downstream; that is the point of topics naming
//  streams, not transports.
//
//      node demo/js/clock.mjs ios | android
//

import { createSuperLog } from '@super-log/client';

const DEVICES = {
  ios: { topic: 'expo.ios.sim', platform: 'ios', device: 'iPhone 16 Pro (sim)' },
  android: { topic: 'expo.android.emu', platform: 'android', device: 'Pixel 8 (emu)' },
};

const kind = DEVICES[process.argv[2]];
if (!kind) {
  console.error('usage: node demo/js/clock.mjs ios|android');
  process.exit(2);
}

const slog = createSuperLog({
  url: process.env.SUPER_LOG_URL ?? 'http://127.0.0.1:7333',
  topic: kind.topic,
  app: 'clock',
  platform: kind.platform,
  device: kind.device,
  development: true, // the bench IS development; a real app passes __DEV__
});

// UTC like the other clients - four streams that agree beat local drift
const hms = () => new Date().toISOString().slice(11, 19) + 'Z';

let n = 0;
slog.info(`${kind.platform} clock up - one line a second`);
const timer = setInterval(() => {
  n += 1;
  slog.info(`tick ${n} - the time is ${hms()}`, { tick: String(n) });
}, 1000);

for (const sig of ['SIGINT', 'SIGTERM']) {
  process.on(sig, () => {
    clearInterval(timer);
    void slog.close().then(() => process.exit(0));
  });
}

File: demo/go/clock.go
// clock.go - the Go demo client
//
// # Copyright 2026 Saxon Herschel Nicholls
//
// The same clock the other demo clients run, once a second on topic
// go.clock, plus the two things a Go service actually wants: the standard
// slog output forwarded with no call-site changes, and a panic in a
// goroutine reaching the bench instead of vanishing.
//
//	go run ./demo/go/clock.go
//	go run ./demo/go/clock.go --crash
package main

import (
	"context"
	"log/slog"
	"os"
	"runtime"
	"time"

	"github.com/saxonnicholls/super-log/sdk/go/superlog"
)

func priceFor(symbol string) (float64, error) {
	rates := map[string]float64{"BTC": 64000, "ETH": 3200}
	v, ok := rates[symbol]
	if !ok {
		return 0, &missingRate{symbol}
	}
	return v, nil
}

type missingRate struct{ symbol string }

func (e *missingRate) Error() string { return "no rate for " + e.symbol }

func main() {
	url := os.Getenv("SUPER_LOG_URL")
	if url == "" {
		url = "http://127.0.0.1:7333"
	}
	log, err := superlog.New(superlog.Config{
		URL: url, Topic: "go.clock", App: "clock", Development: true,
	})
	if err != nil {
		panic(err)
	}
	defer log.Close()

	// Everything already logged through slog reaches the bench, with no
	// changes at any call site. This is the whole point of the handler.
	slog.SetDefault(slog.New(log.SlogHandler(&slog.HandlerOptions{Level: slog.LevelDebug})))

	log.Info("go clock up - one line a second", superlog.F{"go": runtime.Version()})

	if len(os.Args) > 1 && os.Args[1] == "--crash" {
		// A panic in a goroutine is invisible to the main one; Recover puts
		// it on the bench and then re-panics, so the process still dies.
		go func() {
			defer log.Recover("worker")
			var m map[string]int
			m["boom"] = 1 // assignment to nil map
		}()
		time.Sleep(2 * time.Second)
		return
	}

	ticks := 0
	for {
		ticks++
		// One trace per tick, carried by the context - which is how Go
		// passes request-scoped values, so anything this calls inherits it.
		ctx, _ := superlog.WithTrace(context.Background(), "")
		log.LogContext(ctx, superlog.LevelInfo,
			"tick "+time.Now().UTC().Format("15:04:05")+"Z",
			superlog.F{"tick": ticks})
		slog.DebugContext(ctx, "pricing pass", "tick", ticks)

		if ticks%5 == 0 {
			log.Metric("clock.uptime_s", float64(ticks))
		}
		if ticks%7 == 0 {
			if _, err := priceFor("DOGE"); err != nil {
				log.LogContext(ctx, superlog.LevelError, "pricing failed: "+err.Error(),
					superlog.F{"symbol": "DOGE", "tick": ticks})
			}
		}
		time.Sleep(time.Second)
	}
}
